core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java

```diff
@@ -213,10 +206,17 @@ FileIO io(String location) {
               }
             }
           }
+
+          Preconditions.checkState(
```

Comment:
This is fine, but we could also do the cast in the try above  and catch the class cast exception. I'm fine with either approach and think there are benefits to both.

Reply:
I'd slightly prefer the check here as it's more explicit